### PR TITLE
Aggregate changes for large batches

### PR DESCRIPTION
**Pull request by milo-hyben**
This PR contains:
1. Small materialised view change to be bale to see hail batch query jobs
2. Disable monthly aggregate function to populate google doc as this has been replaced with metamist billing report.
3. Aggregate large batches (over 9K jobs per batch - this is set by DEFAULT_MAX_JOBS_PER_BATCH), all jobs billing information gets merged into one job with the JobId of the last jobs, this help us to keep number of jobs per batch.
4. Increase cloud function from 9mins to 1H, there is only one concurrent allowed so if Subscription triggers another after 10mins (max PubSub timeout possible for push), it would fail and wait for the first to complete.

I have tested on Feb 21-Feb 20, which took over 4 hours before to run from local machine. Much faster now.
We can discuss if 9K is not too restrictive, but noticed not many non hail batch query have more than 8K of jobs.

### CODE REVIEW

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:1297]: Why wouldn't we want to FAIL if we can't get the records, and return an empty set - this would mean duplicate records

File: cpg_infra/billing_aggregator/driver.py
Comment (illusional) [inline cpg_infra/billing_aggregator/driver.py:65]: Can you remove all the code for this function to?

```suggestion
```

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:472]: These two limits are for different things, this limit is the _number of jobs to get per request_, not the total number of jobs before we back out. I think you're better to do the query_on_batch check before it gets to here, even write a specific function that takes in a batch_record, and returns True if we guess it to be a hail query on batch, batch (name is blank, and n_jobs > 10k ish).

**Review by illusional — CHANGES_REQUESTED** [review]
Looking good! We did a heap of optimisation so that we don't need the staged ranges (ie: timedelta(day=1)), I'm not opposed to re-adding it for a debug mode, but I think it might be better to be build into the "process_start_end" to produce a list of date ranges [ like it used to be before the optimisation ].

I think the max job limit for detecting hail query isn't quite right, happy to chat if I'm misunderstanding!

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/utils.py]: Yes, makes sense, removed the try/except.

File: cpg_infra/billing_aggregator/driver.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/driver.py]: Done, removed bot call and the function.

File: cpg_infra/billing_aggregator/aggregate/hail.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/hail.py]: As function name suggests *_example() it is just example how to load day by day. I had only used it to load multiple days in a row for testing. 
The __main__ function is example/test as well it is just doing from now to 4 hours prior.
I had removed it.

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/utils.py]: Removed that, taking care of batch queries prior, checking name attribute.

**Comment by milo-hyben** [comment]
Made a few changes, reflecting your comments:
1. Hail Batch Query is identified now as batch with no attribute name
2. Still cut max jobs per batch if jobs_cnt > 9K (happy to discuss if too restrictive)
3. Added option to load batches by batch_id

Please let me know what do you think.

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:630]: - We have batches over the max_jobs_per_batch that we want to process
- We have batches without names under the count we want to process

```suggestion
        if batch_name is None and jobs_cnt > DEFAULT_MAX_JOBS_PER_BATCH:
```

File: cpg_infra/billing_aggregator/aggregate/hail.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/hail.py:214]: The start / end are not used batch_ids flow of `main` - so I don't think there's a need for this.
And if for some reason you do, set them in the `main` function where you get the batches.

**Review by illusional — APPROVED** [review]
Looks good! Two changes that must be addressed before merging, always happy to re-review if you want to double check my suggestions

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/utils.py:630]: Ok, that makes sense, I have thought we want to skip very large ones no matter if HB Query or not. Will change to 'and' so only very large HB Query will be aggregated.

File: cpg_infra/billing_aggregator/aggregate/hail.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/hail.py:214]: Yes, we need the start and end date to retrieve 'stored_ids' from BQ table, prior to calling each batch separately. This was intruded in this PR to optimise those multiple BQ calls with each Hail Batch individually.

**Comment by milo-hyben** [comment]
I had added CI Jobs and for jobs with 0 cost, it adds 0 cost to job resources so it is included in the aggregate table.
I had checked Oct-Dec 2023 and we are missing 1306 jobs, will slack you more details.

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:523]: You don't like this logging?

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:665]: Do you want to call this something notable so it's easier to track down in the future, like '<empty-batch-cost-placeholder>' or will that break things?

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:1139]: Prefer to early-return in Python if practical

```python
if data is None:
    return None

# do other stuff
```

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:513]: Ehhh, it's not a big deal I guess, but I see lots of individual calls to this. I wonder if it's worth adding a cache decorator to this function?

**Review by illusional — APPROVED** [review]
Looking good!

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/utils.py:523]: Was printing out missing batches, so disabled it for my testing. I will put back in.

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/utils.py:665]: Let me try '', makes more sense.

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/utils.py:513]: It is called with different params, so do not think it is wort it. It would be called max 2x per run: 
if loaded by batch id (max 2x), if by date then only once.

File: cpg_infra/billing_aggregator/aggregate/utils.py
Comment (milo-hyben) [reply cpg_infra/billing_aggregator/aggregate/utils.py]: So yes {' ': 0} works.

**Comment by milo-hyben** [comment]
I had resolved the comments. One last review please(can't merge without it). Thank you!

**Review by illusional — APPROVED** [review]

**Pull request merged**
